bigchaindb/util.py

```diff
@@ -22,3 +30,191 @@ def start(self):
+        payload (Optional[dict]): dictionary with information about asset.
+
+    Returns:
+        dict: unsigned transaction.
+
+
+    Raises:
+        TypeError: if the optional ``payload`` argument is not a ``dict``.
+    """
+
+    data = None
+    if payload is not None:
+        if isinstance(payload, dict):
+            hash_payload = hash_data(serialize(payload))
+            data = {
+                'hash': hash_payload,
+                'payload': payload
+            }
+        else:
+            raise TypeError('`payload` must be an dict instance')
+
+    hash_payload = hash_data(serialize(payload))
+    data = {
+        'hash': hash_payload,
+        'payload': payload
+    }
+
+    tx = {
+        'current_owner': current_owner,
+        'new_owner': new_owner,
+        'input': tx_input,
+        'operation': operation,
+        'timestamp': timestamp(),
+        'data': data
+    }
+
+    # serialize and convert to bytes
+    tx_serialized = serialize(tx)
+    tx_hash = hash_data(tx_serialized)
+
+    # create the transaction
+    transaction = {
+        'id': tx_hash,
+        'transaction': tx
+    }
+
+    return transaction
+
+
+def sign_tx(transaction, private_key):
+    """Sign a transaction
+
+    A transaction signed with the `current_owner` corresponding private key.
+
+    Args:
+        transaction (dict): transaction to sign.
+        private_key (str): base58 encoded private key to create a signature of the transaction.
+
+    Returns:
+        dict: transaction with the `signature` field included.
+
+    """
+    private_key = PrivateKey(private_key)
+    signature = private_key.sign(serialize(transaction))
+    signed_transaction = transaction.copy()
+    signed_transaction.update({'signature': signature})
+    return signed_transaction
+
+
+def create_and_sign_tx(private_key, current_owner, new_owner, tx_input, operation='TRANSFER', payload=None):
+    tx = create_tx(current_owner, new_owner, tx_input, operation, payload)
+    return sign_tx(tx, private_key)
+
+
+def check_hash_and_signature(transaction):
+    # Check hash of the transaction
+    calculated_hash = hash_data(serialize(transaction['transaction']))
+    if calculated_hash != transaction['id']:
+        raise exceptions.InvalidHash()
+
+    # Check signature
+    if not verify_signature(transaction):
+        raise exceptions.InvalidSignature()
+
+
+def verify_signature(signed_transaction):
+    """Verify the signature of a transaction
+
+    A valid transaction should have been signed `current_owner` corresponding private key.
+
+    Args:
+        signed_transaction (dict): a transaction with the `signature` included.
+
+    Returns:
+        bool: True if the signature is correct, False otherwise.
+    """
+
+    data = signed_transaction.copy()
+
+    # if assignee field in the transaction, remove it
+    if 'assignee' in data:
+        data.pop('assignee')
+
+    signature = data.pop('signature')
```

Comment:
Yup, because it's required, and it's totally fine. Data validation must be done **before** calling the function.

If you pass a transaction without `signature`, then it's an **error**, because there is no signature to verify, and the function will raise a `KeyError`.